src/core/compact_object.cc

```diff
@@ -138,39 +138,74 @@ inline void FreeObjZset(unsigned encoding, void* ptr) {
   }
 }
 
+pair<void*, bool> DefragStrMap2(StringMap* sm, float ratio) {
+  bool realloced = false;
+
+  for (auto it = sm->begin(); it != sm->end(); ++it)
+    realloced |= it.ReallocIfNeeded(ratio);
+
+  return {sm, realloced};
+}
+
+pair<void*, bool> DefragListPack(uint8_t* lp, float ratio) {
+  if (!zmalloc_page_is_underutilized(lp, ratio))
+    return {lp, false};
+
+  size_t lp_bytes = lpBytes(lp);
+  uint8_t* replacement = lpNew(lpBytes(lp));
+  memcpy(replacement, lp, lp_bytes);
+  lpFree(lp);
+
+  return {replacement, true};
+}
+
+pair<void*, bool> DefragIntSet(intset* is, float ratio) {
+  if (!zmalloc_page_is_underutilized(is, ratio))
+    return {is, false};
+
+  const size_t blob_len = intsetBlobLen(is);
+  intset* replacement = (intset*)zmalloc(blob_len);
+  memcpy(replacement, is, blob_len);
```

Comment:
This is based on my understanding of how `intset` is implemented. I could be wrong. 

1. I can test this locally that memcpy works so we don't run into trouble
2. Are there any tests that hit this path @adiholden